allennlp/interpret/attackers/hotflip.py

```diff
@@ -0,0 +1,184 @@
+
+            current_input_field = new_instances[0][input_field_to_attack]
+            current_tokens = getattr(current_input_field, 'tokens')
+            grads, outputs = self.predictor.get_gradients(new_instances)
+            flipped = []  # type: List[int]
+            while True:
+                # Compute L2 norm of all grads.
+                grad = grads[grad_input_field]
+                grads_magnitude = [g.dot(g) for g in grad]
+
+                # only flip a token once
+                for index in flipped:
+                    grads_magnitude[index] = -1
+
+                # we flip the token with highest gradient norm
+                index_of_token_to_flip = numpy.argmax(grads_magnitude)
+                # when we have already flipped all the tokens once
+                if grads_magnitude[index_of_token_to_flip] == -1:
+                    break
+                flipped.append(index_of_token_to_flip)
+
+                # Get new token using taylor approximation
+                input_tokens = new_instances[0][input_field_to_attack]._indexed_tokens["tokens"] # type: ignore
+                original_id_of_token_to_flip = input_tokens[index_of_token_to_flip]
+                new_id_of_flipped_token = first_order_taylor(grad[index_of_token_to_flip],
+                                                             self.token_embedding.weight,
+                                                             original_id_of_token_to_flip)
+                # flip token
+                new_token = Token(self.vocab._index_to_token["tokens"][new_id_of_flipped_token]) # type: ignore
+                new_instances[0][input_field_to_attack].tokens[index_of_token_to_flip] = new_token # type: ignore
+                new_instances[0].indexed = False
+
+                # Get model predictions on new_instances, and then label the instances
+                grads, outputs = self.predictor.get_gradients(new_instances) # predictions
+                for key in outputs:
+                    if isinstance(outputs[key], torch.Tensor):
+                        outputs[key] = outputs[key].detach().cpu().numpy().squeeze()
+                    elif isinstance(outputs[key], list):
+                        outputs[key] = outputs[key][0]
+                # add labels to new_instances
+                self.predictor.predictions_to_labeled_instances(new_instances[0], outputs)
+
+                # if the prediction has changed, then stop
+                label_change = False
+                for field in fields_to_compare:
+                    if field in new_instances[0].fields:
+                        equal = new_instances[0][field] == fields_to_compare[field]
+                    else:
+                        equal = outputs[field] == fields_to_compare[field]
+                    if not equal:
+                        label_change = True
+                        break
+                if label_change:
+                    break
```

Comment:
"flat is better than nested", there must be a clearer way to accomplish this, something like

```
def is_equal(field: str) -> bool:
    if field in new_instances[0].fields:
        return ... == ...
    else:
        return ... == ...

if any(is_equal(field) for field in fields_to_compare): 
    break
```

but also the logic for "the prediction has changed" is really not obvious to me, is there a way to make it clearer